File: mobile/android/app/src/main/java/com/tigerwallet/app/trading/CryptoCardService.java
package com.tigerwallet.app.trading;

/**
 * Crypto Card Service - Android Implementation
 * Virtual and Physical Crypto Cards
 */

import java.util.UUID;
import java.util.Random;

public class CryptoCardService {
    private Random random = new Random();
    
    public static class CryptoCard {
        public String id;
        public String userId;
        public String cardNumber;
        public String cardHolder;
        public String expiryDate;
        public String cvv;
        public String type;
        public String network;
        public String status;
        public double dailyLimit;
        public double monthlyLimit;
        public double dailySpent;
        public double monthlySpent;
        public boolean applePayEnabled;
        public boolean googlePayEnabled;
        public String maskedNumber;
    }
    
    public static class CardTransaction {
        public String id;
        public String cardId;
        public String userId;
        public double amount;
        public String currency;
        public String merchantName;
        public String status;
        public long timestamp;
    }
    
    private String generateCardNumber() {
        return "4532" + String.format("%012d", Math.abs(random.nextLong()) % 1000000000000L);
    }
    
    private String generateCVV() {
        return String.format("%03d", random.nextInt(900) + 100);
    }
    
    private String generateExpiry() {
        int month = random.nextInt(12) + 1;
        int year = (2024 + 3) % 100;
        return String.format("%02d/%02d", month, year);
    }
    
    public CryptoCard createVirtualCard(String userId, String cardHolder, String type, String network) {
        CryptoCard card = new CryptoCard();
        card.id = "card_" + UUID.randomUUID().toString();
        card.userId = userId;
        card.cardNumber = generateCardNumber();
        card.cardHolder = cardHolder;
        card.expiryDate = generateExpiry();
        card.cvv = generateCVV();
        card.type = type != null ? type : "VIRTUAL";
        card.network = network != null ? network : "VISA";
        card.status = "ACTIVE";
        card.dailyLimit = 10000;
        card.monthlyLimit = 100000;
        card.dailySpent = 0;
        card.monthlySpent = 0;
        card.applePayEnabled = true;
        card.googlePayEnabled = true;
        card.maskedNumber = "•••• •••• •••• " + card.cardNumber.substring(card.cardNumber.length() - 4);
        return card;
    }
    
    public CryptoCard createPhysicalCard(String userId, String cardHolder, String shippingAddress) {
        CryptoCard card = createVirtualCard(userId, cardHolder, "PHYSICAL", "VISA");
        card.status = "PENDING_ACTIVATION";
        return card;
    }
    
    public CardTransaction processPayment(CryptoCard card, double amount, String currency, String merchantName) {
        if (!"ACTIVE".equals(card.status)) {
            throw new RuntimeException("Card is not active");
        }
        if (card.dailySpent + amount > card.dailyLimit) {
            throw new RuntimeException("Daily limit exceeded");
        }
        
        card.dailySpent += amount;
        card.monthlySpent += amount;
        
        CardTransaction txn = new CardTransaction();
        txn.id = "txn_" + System.currentTimeMillis();
        txn.cardId = card.id;
        txn.userId = card.userId;
        txn.amount = amount;
        txn.currency = currency;
        txn.merchantName = merchantName;
        txn.status = "COMPLETED";
        txn.timestamp = System.currentTimeMillis();
        
        return txn;
    }
    
    public void freezeCard(CryptoCard card) {
        card.status = "FROZEN";
    }
    
    public void unfreezeCard(CryptoCard card) {
        card.status = "ACTIVE";
    }
    
    public void terminateCard(CryptoCard card) {
        card.status = "TERMINATED";
        card.applePayEnabled = false;
        card.googlePayEnabled = false;
    }
    
    public void enableApplePay(CryptoCard card) {
        card.applePayEnabled = true;
    }
    
    public void enableGooglePay(CryptoCard card) {
        card.googlePayEnabled = true;
    }
}

File: mobile_apps/ios_legacy/TigerWallet/Services/Trading/CryptoCardService.swift
//
//  CryptoCardService.swift
//  TigerWallet
//
//  Crypto Card Service - iOS Implementation
//

import Foundation

struct CryptoCard {
    var id: String
    var userId: String
    var cardNumber: String
    var cardHolder: String
    var expiryDate: String
    var cvv: String
    var type: String
    var network: String
    var status: String
    var dailyLimit: Double
    var monthlyLimit: Double
    var dailySpent: Double
    var monthlySpent: Double
    var applePayEnabled: Bool
    var googlePayEnabled: Bool
    var maskedNumber: String
}

struct CardTransaction {
    var id: String
    var cardId: String
    var userId: String
    var amount: Double
    var currency: String
    var merchantName: String
    var status: String
    var timestamp: Date
}

class CryptoCardService {
    
    func generateCardNumber() -> String {
        let random = String(format: "%012d", Int.random(in: 0...999999999999))
        return "4532" + random
    }
    
    func generateCVV() -> String {
        return String(format: "%03d", Int.random(in: 100...999))
    }
    
    func generateExpiry() -> String {
        let month = Int.random(in: 1...12)
        let year = (Calendar.current.component(.year, from: Date()) + 3) % 100
        return String(format: "%02d/%02d", month, year)
    }
    
    func createVirtualCard(userId: String, cardHolder: String, type: String? = "VIRTUAL", network: String? = "VISA") -> CryptoCard {
        let cardNumber = generateCardNumber()
        return CryptoCard(
            id: "card_\(Int(Date().timeIntervalSince1970 * 1000))",
            userId: userId,
            cardNumber: cardNumber,
            cardHolder: cardHolder,
            expiryDate: generateExpiry(),
            cvv: generateCVV(),
            type: type ?? "VIRTUAL",
            network: network ?? "VISA",
            status: "ACTIVE",
            dailyLimit: 10000,
            monthlyLimit: 100000,
            dailySpent: 0,
            monthlySpent: 0,
            applePayEnabled: true,
            googlePayEnabled: true,
            maskedNumber: "•••• •••• •••• " + String(cardNumber.suffix(4))
        )
    }
    
    func createPhysicalCard(userId: String, cardHolder: String, shippingAddress: String) -> CryptoCard {
        var card = createVirtualCard(userId: userId, cardHolder: cardHolder, type: "PHYSICAL")
        card.status = "PENDING_ACTIVATION"
        return card
    }
    
    func processPayment(card: inout CryptoCard, amount: Double, currency: String, merchantName: String) throws -> CardTransaction {
        guard card.status == "ACTIVE" else {
            throw NSError(domain: "CryptoCard", code: 1, userInfo: [NSLocalizedDescriptionKey: "Card is not active"])
        }
        
        guard card.dailySpent + amount <= card.dailyLimit else {
            throw NSError(domain: "CryptoCard", code: 2, userInfo: [NSLocalizedDescriptionKey: "Daily limit exceeded"])
        }
        
        card.dailySpent += amount
        card.monthlySpent += amount
        
        return CardTransaction(
            id: "txn_\(Int(Date().timeIntervalSince1970 * 1000))",
            cardId: card.id,
            userId: card.userId,
            amount: amount,
            currency: currency,
            merchantName: merchantName,
            status: "COMPLETED",
            timestamp: Date()
        )
    }
    
    func freezeCard(card: inout CryptoCard) {
        card.status = "FROZEN"
    }
    
    func unfreezeCard(card: inout CryptoCard) {
        card.status = "ACTIVE"
    }
    
    func terminateCard(card: inout CryptoCard) {
        card.status = "TERMINATED"
        card.applePayEnabled = false
        card.googlePayEnabled = false
    }
    
    func enableApplePay(card: inout CryptoCard) {
        card.applePayEnabled = true
    }
    
    func enableGooglePay(card: inout CryptoCard) {
        card.googlePayEnabled = true
    }
}
